File: app/src/main/java/com/leaguechampions/ui/champions/ChampionsPresenter.java
package com.leaguechampions.ui.champions;

import android.os.Bundle;
import android.support.annotation.StringRes;
import android.view.MenuItem;

import com.leaguechampions.R;
import com.leaguechampions.data.model.Champion;
import com.leaguechampions.data.model.RiotResponse;
import com.leaguechampions.data.remote.Api;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class ChampionsPresenter implements ChampionsAdapter.onItemClickListener {

    private final Api api;
    private final ChampionsView view;

    public interface ChampionsView {
        void setAdapter(List<Champion> champions);
        void showError(@StringRes int stringId);
        void showDetails(String championId);
        void showSettings();
    }

    @Inject
    public ChampionsPresenter(ChampionsView view, Api api) {
        this.view = view;
        this.api = api;
    }

    public void onActivityCreated(Bundle savedInstanceState, Bundle arguments) {
        getChampions();
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_settings:
                view.showSettings();
                return true;
            default:
                return false;
        }
    }

    @Override
    public void onItemClick(Champion champion) {
        view.showDetails(champion.getId());
    }

    private void getChampions() {
        api.getVersion()
                .flatMap(response ->
                        api.getChampions(response.getVersion())
                )
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<RiotResponse>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(RiotResponse response) {
                        List<Champion> champions = new ArrayList<>(response.getData().values());
                        Collections.sort(champions);
                        view.setAdapter(champions);
                    }

                    @Override
                    public void onError(Throwable t) {
                        Timber.e(t);
                        if (t instanceof IOException) {
                            view.showError(R.string.error_io);
                        } else {
                            view.showError(R.string.error_something_went_wrong);
                        }
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }
}

File: app/src/main/java/com/leaguechampions/ui/champions/ChampionsPresenter.kt
package com.leaguechampions.ui.champions

import android.os.Bundle
import android.support.annotation.StringRes
import android.view.MenuItem
import com.leaguechampions.R
import com.leaguechampions.data.model.Champion
import com.leaguechampions.data.model.RiotResponse
import com.leaguechampions.data.remote.Api
import javax.inject.Inject

class ChampionsPresenter @Inject constructor(private val view: ChampionsView, private val api: Api)
//    : ChampionsAdapter.OnItemClickListener
{

    interface ChampionsView {
        fun setAdapter(riotResponse: RiotResponse)
        fun showError(@StringRes stringId: Int)
        fun showError(@StringRes stringId: Int, errorCode: Int)
        fun showDetails(version: String, championId: String)
        fun showSettings()
    }

    fun onActivityCreated(savedInstanceState: Bundle?, arguments: Bundle?) {
        getChampions()
    }

    fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_settings -> {
                view.showSettings()
                true
            }
            else -> false
        }
    }

//    override
    fun onItemClick(version: String, champion: Champion) {
        view.showDetails(version, champion.id)
    }

    private fun getChampions() {
//        api.getChampions().enqueue(object : Callback<RiotResponse> {
//            override fun onResponse(call: Call<RiotResponse>, response: Response<RiotResponse>) {
//                if (response.isSuccessful) {
//                    view.setAdapter(response.body()!!)
//                } else {
//                    view.showError(R.string.error_code, response.code())
//                }
//            }
//
//            override fun onFailure(call: Call<RiotResponse>, t: Throwable) {
//                view.showError(if (t is IOException) R.string.error_io else R.string.error_something_went_wrong)
//            }
//        })
    }
}
